### ISSUE TITLE
/* comments */ should not be reformatted

### ISSUE THREAD

**Issue by toastal** [NixOS/nixfmt#179]
### Description

`/* */` style comments should not be reformatted. They have additional use cases such as tree-sitter picking up the the syntax to highlight before scripts inside multiline strings.

### Small example input

```nix
/* lua */ ''
  print("Hello, world!")
''
```

### Expected output

```nix
/* lua */ ''
  print("Hello, world!")
''
```

(now tree-sitter in my editor highlights this string to Lua which really helps me read/write code in this block now that it’s not just syntax highlighted as a string)

### Actual output

```nix
# lua
''
  print("Hello, world!")
''
```

**Comment by dasJ** [comment]
Fwiw, nvim-treesitter now supports that: https://github.com/nvim-treesitter/nvim-treesitter/pull/6418#issuecomment-2044072426

**Comment by SuperSandro2000** [comment]
> They have additional use cases such as tree-sitter picking up the the syntax to highlight before scripts inside multiline strings.

or fenced vim hints.

> nvim-treesitter now supports that

That's not released, yet.

**Comment by infinisil** [comment]
We discussed this in the team meeting today:

  - Sounds okay, but a bit ad-hoc, would be good to have a standard for such annotations
    - @tomberek: variations: `/*!` `/*nofmt` `/*^` `/*lang:lua` 
  - @tomberek: Can't upstream treesitter also support the `# <lang>` style comments?
    - @infinisil: Should be possible!
    - @das_j: Is now possible for at least nvim-treesitter: https://github.com/nvim-treesitter/nvim-treesitter/pull/6418#pullrequestreview-1988027877, but might not be used by downstream projects
    - @infinisil: Seems like it's an easy fix, a similar patch can be used for other treesitter Nix injection definitions
  - Original motivations:
    - Worked best with the implementation, hard to change
    - Didn't know a use case
    - What to do if there's code after a `/* */` single-line comment?
      - We can insert a newline, e.g. from
        ```
        /* long comment */ foo + 10
        ```
        to
        ```
        /* long comment */
        foo + 10
        ```
    - Previously they were often used for vertical alignment

Conclusion: We generally think that it would be fine to preserve `/*`-style comments more. While we don't intend to fix this ourselves, anybody is free to PR this. If you do, make sure to update [this line](https://github.com/NixOS/nixfmt/blob/87c4879b7a72a69726b76392847fb79a7044c6a6/standard.md#L398) in the standard.

**Comment by piegamesde** [comment]
The issue is less with the parsing, that's a bit fiddly but no more than a chore. The big open question is, what to do in the renderer with such comments when they are followed by significant amounts of code which overflow the line length limit.

The rendering engine currently ignores all line length calculations involving comments, however this strongly depends on the assumption that comment tokens are always strictly followed by a line break.

**Comment by MattSturgeon** [comment]
Thanks for clarifying!

> The big open question is, what to do in the renderer with such comments when they are followed by significant amounts of code which overflow the line length limit.

Ideally, inserting a line break _before_ the comment would be enough to not overflow, but that's the best case scenario...

I think it's fair to say, that if someone has written an inline block comment on the same line as a long string (for example), it was intentional. Maybe in that scenario it is just accepted that the line will overflow?\
Alternatively, perhaps a `''` indented raw string could be used, with the string content starting after a `"''\n""` sequence? Though this seems overly complex at first glance.

If the long content to the right of the inline block comment is not a string, then I think it is ok to reformat that as usual, such that it is wrapped over multiple lines.

> The rendering engine currently ignores all line length calculations involving comments, however this strongly depends on the assumption that comment tokens are always strictly followed by a line break.

That sounds like a bigger issue. Perhaps "inline" comments can be a distinct type from other comments and therefore included in line length calculations?

**Comment by piegamesde** [comment]
So one thing that one could do would be to special case these comments in the parser but only if they are followed by `''` and then a line break. I think that should cover all reasonable use cases without breaking everything else.

**Comment by MattSturgeon** [comment]
> So one thing that one could do would be to special case these comments in the parser but only if they are followed by `''` and then a line break. I think that should cover all reasonable use cases without breaking everything else.

While it'd be _nice_ to have support for normal strings too, I agree that is a reasonable approach that'd work for most usage.

I do sometimes do things like this in my config.

```nix
{
  # This snippet is so short it feels strange to use an indented string:
  someLuaOption = /* lua */ "function() print('hello, world!') end";
}
```

But some support is better than no support, and it can always be iterated on in the future.

**Comment by ian-h-chamberlain** [comment]
Just to propose one additional use case for `/* */` comment (non-)formatting: it's possible to toggle a block comment with a single `#` using something like the following:
```nix
{
  /* <-- Toggling a # line comment at the beginning of this line toggles the whole block
    services.foo = {
      enable = true;
      bar = "baz";
    };
  # */
}
```

For this kind of case I think it would be valuable to avoid splitting the final `# */` onto two lines (and maybe also the first line, but that's less important). Otherwise this "quick toggle" results in invalid syntax, e.g.
```nix
{
  # /*
    <-- Toggling a # line comment at the beginning of this line toggles the whole block
      services.foo = {
        enable = true;
        bar = "baz";
      };
    #
  */
}
```

I use this in my configs to quickly toggle on/off large sections of the config. I'm not sure how popular a practice it is in general, but it seems to be moderately common from a cursory search: 
https://github.com/search?q=lang%3Anix+%2F%28%23+%5C*%5C%2F%7C%23+%5C%2F%5C*%29%2F&type=code

**Comment by MattSturgeon** [comment]
> Just to propose one additional use case for `/* */` comment (non-)formatting

I don't think this issue is proposing that block comments not be formatted; instead this issue is about allowing _short_ and/or _inline_ `/* */` comments.

Given nixfmt is deterministic, it'd probably have to be based on how long the comment content is and whether or not there is trailing code after the comment?

Your issue about block comments being formatted messing up your workflow should probably be tracked in a separate ticket and may be best implemented as a configurable cli flag.

**Comment by piegamesde** [comment]
@ian-h-chamberlain commenting out code blocks with /* has the advantage that it is low on diff noise, however it has the big disadvantage that it breaks as soon as that block contains another nested /* comment.

**Comment by ian-h-chamberlain** [comment]
I opened #225 to track separately, maybe it would also make sense to update the title of this issue to be more specifically about short/inline `/* */` comments just to clarify? 
Any more discussion about whether / how to handle this case can continue on that issue, I suppose.

**Comment by infinisil** [comment]
Btw here's the code that handles the conversion from `/* foo */` to `# foo`: https://github.com/NixOS/nixfmt/blob/e819b2d0f9173f0c73d2e1fb4bc5155362046653/src/Nixfmt/Lexer.hs#L131-L169

**Comment by omkumar312** [comment]
I want to work on this issues please assign me this issue.

**Comment by SuperSandro2000** [comment]
> Btw here's the code that handles the conversion from `/* foo */` to `# foo`:

Does that mean we just need to drop that code and we are good to go?

**Comment by Lau-San** [comment]
Any updates on this?

**Comment by jfly** [comment]
@Lau-San, we forgot to discuss this in today's team meeting, sorry. I've added this to our list of things to discuss next time.
